File: python/sim_mini/telemetry.py
"""64-byte telemetry frame codec.

Mirrors ``packages/sim/src/telemetry.ts`` byte-for-byte.
"""

from dataclasses import dataclass
from hashlib import sha256

FRAME_BYTES = 64
PAYLOAD_BYTES = 32
HEADER_BYTES = 48
SHA_PREFIX_BYTES = 8


@dataclass(frozen=True)
class TelemetryFrame:
    sync: int
    version: int
    seq: int
    t_us: int
    layer: int
    kind: int
    flags: int
    payload: bytes
    sha256_prefix: bytes


def _sha256_prefix(buf: bytes) -> bytes:
    return sha256(buf).digest()[:SHA_PREFIX_BYTES]


def _assert_frame_buffer(buf: bytes) -> None:
    if len(buf) != FRAME_BYTES:
        raise ValueError("frame must be 64 bytes")


def _payload_to_bytes(payload: bytes) -> bytes:
    if len(payload) != PAYLOAD_BYTES:
        raise ValueError("payload must be 32 bytes")
    return payload


def create_telemetry_frame(
    *,
    seq: int,
    t_us: int,
    layer: int,
    kind: int,
    flags: int,
    payload: bytes,
    sha256_prefix: bytes | None = None,
) -> TelemetryFrame:
    provisional = bytearray(FRAME_BYTES)
    provisional[0] = 0xCB
    provisional[1] = 0x01
    provisional[2:4] = seq.to_bytes(2, "little", signed=False)
    provisional[4:12] = t_us.to_bytes(8, "little", signed=False)
    provisional[12] = layer
    provisional[13] = kind
    provisional[14:16] = flags.to_bytes(2, "little", signed=False)
    provisional[16:48] = _payload_to_bytes(payload)
    prefix = (
        sha256_prefix
        if sha256_prefix is not None
        else _sha256_prefix(bytes(provisional[:HEADER_BYTES]))
    )

    return TelemetryFrame(
        sync=0xCB,
        version=1,
        seq=seq,
        t_us=t_us,
        layer=layer,
        kind=kind,
        flags=flags,
        payload=payload,
        sha256_prefix=prefix,
    )


def encode_telemetry_frame(frame: TelemetryFrame) -> bytes:
    payload = _payload_to_bytes(frame.payload)

    encoded = bytearray(FRAME_BYTES)
    encoded[0] = frame.sync
    encoded[1] = frame.version
    encoded[2:4] = frame.seq.to_bytes(2, "little", signed=False)
    encoded[4:12] = frame.t_us.to_bytes(8, "little", signed=False)
    encoded[12] = frame.layer
    encoded[13] = frame.kind
    encoded[14:16] = frame.flags.to_bytes(2, "little", signed=False)
    encoded[16:48] = payload
    if len(frame.sha256_prefix) != SHA_PREFIX_BYTES:
        raise ValueError("sha256_prefix does not match bytes 0..47")
    encoded[48:56] = frame.sha256_prefix
    encoded[56:64] = b"\x00" * 8

    if _sha256_prefix(bytes(encoded[:HEADER_BYTES])) != frame.sha256_prefix:
        raise ValueError("sha256_prefix does not match bytes 0..47")

    return bytes(encoded)


def decode_telemetry_frame(buf: bytes) -> TelemetryFrame:
    _assert_frame_buffer(buf)
    if buf[0] != 0xCB:
        raise ValueError("bad sync byte")
    if buf[1] != 0x01:
        raise ValueError("unsupported protocol version")
    if any(buf[56:64]):
        raise ValueError("reserved bytes must be zero")

    expected_prefix = _sha256_prefix(buf[:HEADER_BYTES])
    actual_prefix = buf[48:56]
    if expected_prefix != actual_prefix:
        raise ValueError("sha256_prefix mismatch")

    return TelemetryFrame(
        sync=buf[0],
        version=buf[1],
        seq=int.from_bytes(buf[2:4], "little", signed=False),
        t_us=int.from_bytes(buf[4:12], "little", signed=False),
        layer=buf[12],
        kind=buf[13],
        flags=int.from_bytes(buf[14:16], "little", signed=False),
        payload=buf[16:48],
        sha256_prefix=actual_prefix,
    )

File: packages/sim/src/telemetry.ts
import { createHash } from 'node:crypto';
import type { TelemetryFrame } from './types.js';

const FRAME_BYTES = 64;
const PAYLOAD_BYTES = 32;
const HEADER_BYTES = 48;
const SHA_PREFIX_BYTES = 8;

function assertFrameBuffer(buf: Uint8Array): void {
  if (buf.length !== FRAME_BYTES) {
    throw new Error('frame must be 64 bytes');
  }
}

function readU16LE(buf: Uint8Array, offset: number): number {
  return buf[offset]! | (buf[offset + 1]! << 8);
}

function readU64LE(buf: Uint8Array, offset: number): number {
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  return Number(view.getBigUint64(offset, true));
}

function sha256Prefix(buf: Uint8Array): Uint8Array {
  return createHash('sha256').update(buf).digest().subarray(0, SHA_PREFIX_BYTES);
}

function payloadToBytes(payload: Uint8Array): Uint8Array {
  if (payload.length !== PAYLOAD_BYTES) {
    throw new Error('payload must be 32 bytes');
  }
  return payload;
}

export function encodeTelemetryFrame(frame: TelemetryFrame): Uint8Array {
  const bytes = new Uint8Array(FRAME_BYTES);
  const view = new DataView(bytes.buffer);

  bytes[0] = frame.sync;
  bytes[1] = frame.version;
  view.setUint16(2, frame.seq, true);
  view.setBigUint64(4, BigInt(frame.t_us), true);
  bytes[12] = frame.layer;
  bytes[13] = frame.kind;
  view.setUint16(14, frame.flags, true);
  bytes.set(payloadToBytes(frame.payload), 16);
  bytes.set(frame.sha256_prefix, 48);
  bytes.fill(0, 56);

  const expectedPrefix = sha256Prefix(bytes.subarray(0, HEADER_BYTES));
  if (
    frame.sha256_prefix.length !== SHA_PREFIX_BYTES ||
    !expectedPrefix.every((byte, index) => byte === frame.sha256_prefix[index])
  ) {
    throw new Error('sha256_prefix does not match bytes 0..47');
  }

  return bytes;
}

export function decodeTelemetryFrame(buf: Uint8Array): TelemetryFrame {
  assertFrameBuffer(buf);
  if (buf[0] !== 0xcb) throw new Error('bad sync byte');
  if (buf[1] !== 0x01) throw new Error('unsupported protocol version');
  if (!buf.subarray(56).every((byte) => byte === 0)) {
    throw new Error('reserved bytes must be zero');
  }

  const expectedPrefix = sha256Prefix(buf.subarray(0, HEADER_BYTES));
  const actualPrefix = buf.subarray(48, 56);
  if (!expectedPrefix.every((byte, index) => byte === actualPrefix[index])) {
    throw new Error('sha256_prefix mismatch');
  }

  return {
    sync: buf[0] as 0xcb,
    version: buf[1] as 1,
    seq: readU16LE(buf, 2),
    t_us: readU64LE(buf, 4),
    layer: buf[12] as 1 | 2 | 3 | 4 | 5,
    kind: buf[13]!,
    flags: readU16LE(buf, 14),
    payload: buf.slice(16, 48),
    sha256_prefix: buf.slice(48, 56),
  };
}

export function createTelemetryFrame(
  frame: Omit<TelemetryFrame, 'sync' | 'version' | 'sha256_prefix'> & {
    sha256_prefix?: Uint8Array;
  },
): TelemetryFrame {
  const provisional = new Uint8Array(FRAME_BYTES);
  provisional[0] = 0xcb;
  provisional[1] = 0x01;
  const view = new DataView(provisional.buffer);
  view.setUint16(2, frame.seq, true);
  view.setBigUint64(4, BigInt(frame.t_us), true);
  provisional[12] = frame.layer;
  provisional[13] = frame.kind;
  view.setUint16(14, frame.flags, true);
  provisional.set(payloadToBytes(frame.payload), 16);
  const prefix = frame.sha256_prefix ?? sha256Prefix(provisional.subarray(0, HEADER_BYTES));

  return {
    sync: 0xcb,
    version: 1,
    seq: frame.seq,
    t_us: frame.t_us,
    layer: frame.layer,
    kind: frame.kind,
    flags: frame.flags,
    payload: frame.payload,
    sha256_prefix: prefix,
  };
}
